presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1199,12 +1199,16 @@ private PlanWithProperties withDerivedProperties(PlanNode node, ActualProperties
 
         private ActualProperties deriveProperties(PlanNode result, ActualProperties inputProperties)
         {
-            return PropertyDerivations.deriveProperties(result, inputProperties, metadata, session, types, parser);
+            return deriveProperties(result, ImmutableList.of(inputProperties));
         }
 
         private ActualProperties deriveProperties(PlanNode result, List<ActualProperties> inputProperties)
         {
-            return PropertyDerivations.deriveProperties(result, inputProperties, metadata, session, types, parser);
+            // TODO: move this logic to PlanSanityChecker once PropertyDerivations.deriveProperties fully supports local exchanges
+            ActualProperties outputProperties = PropertyDerivations.deriveProperties(result, inputProperties, metadata, session, types, parser);
+            verify(result instanceof SemiJoinNode || inputProperties.stream().noneMatch(ActualProperties::isNullsAndAnyReplicated) || outputProperties.isNullsAndAnyReplicated(),
```

Comment:
Should we check that above all the `SemiJoins` nulls-and-any are actully **not** replicated in the plan. Or do we do that somewhere already?

Reply:
I don't understand. Don't we check that exactly here?

Reply:
Theoretically output of root of the plan could be null-and-any replicated and it would be allowed by the validator.

Probalby not important.